podman/tests/integration/utils.py

```diff
@@ -0,0 +1,98 @@
+#   Copyright 2020 Red Hat, Inc.
+#
+#   Licensed under the Apache License, Version 2.0 (the "License"); you may
+#   not use this file except in compliance with the License. You may obtain
+#   a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#   WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#   License for the specific language governing permissions and limitations
+#   under the License.
+#
+"""Integration Test Utils"""
+
+import os
+import shutil
+import subprocess
+import tempfile
+import time
+from typing import List, Optional
+
+import podman.tests.errors as errors
+
+
+class PodmanLauncher:
+    """Podman service launcher"""
+
+    def __init__(
+        self,
+        socket_uri: str,
+        podman_path: Optional[str] = None,
+        timeout: int = 0,
+        privileged: bool = False,
+        debug: bool = False,
+    ) -> None:
+        """create a launcher and build podman command"""
+        podman_exe: str = podman_path
+        if not podman_exe:
+            podman_exe = shutil.which('podman')
+        if podman_exe is None:
+            raise errors.PodmanNotInstalled()
+
+        self.socket_file: str = socket_uri.replace('unix://', '')
+        self.proc = None
+
+        self.cmd: List[str] = []
+        if privileged:
+            self.cmd.append('sudo')
+
+        self.cmd.append(podman_exe)
+
+        if debug:
+            self.cmd.append("--log-level=debug")
+
+        self.cmd.extend(
+            [
+                "system",
+                "service",
+                f"--time={timeout}",
+                socket_uri,
+            ]
+        )
+
+        self.log_file = tempfile.NamedTemporaryFile(
+            prefix="podman_integration_", suffix=".log", delete=False
+        )
+        self.log_file.write(f"Podman command {' '.join(self.cmd)}\n".encode())
+        self.log_file.flush()
+
+    def start(self) -> None:
+        """start podman service"""
+        print(f"Launching {' '.join(self.cmd)}\n\tLogging to {self.log_file.name}")
+        self.proc = subprocess.Popen(self.cmd, stdout=self.log_file, stderr=self.log_file)
+
+        # wait for socket to be created
+        timeout = time.monotonic() + 30
+        while not os.path.exists(self.socket_file):
+            if time.monotonic() > timeout:
+                raise subprocess.TimeoutExpired("podman service ", timeout)
+            time.sleep(0.2)
```

Comment:
Is there some positive indicator that could be waited upon instead of sleeping?  Two tenths could either be an eternity on a busy CI system, potentially obscuring another/worse problem.

Reply:
It's waiting for the socket file to be available. I found that even on local systems a bit of delay occurred prior to the service being ready to accept connections.  While you could lower the time to like 0.01, you definitely don't want to let it spin uncontrolled as it'll just eat CPU.

Reply:
Should we have a max tries that would fail the test after 3 or so tries?

Reply:
I wouldn't put a loop cap for the exact reasons cevich mentioned, in CI this could be a while so if you looped 3 times you'd only have waited for ~.6 seconds.  There is a cap here of 30 seconds with the timeout var.  30 seconds should be plenty of time to wait for the podman api instance to become available for the integration tests.

Reply:
@mwhahaha is waiting on the socket to be created on the filesystem. I added the timeout so we don't wait forever and the small sleep should help with CPU usage.

Reply:
Okay, I think I understand what's happening here.  The sleep is to prevent checking for the existence of the file, from happening repeatedly as fast as possible.  I don't see it in the code (I didn't look) but a max of 30-seconds for this loop is probably fine.